rangeserver/src/cache/memtabledb.rs

```diff
@@ -0,0 +1,643 @@
+use skiplist::OrderedSkipList;
+use std::cmp::Ordering;
+use tokio::sync::RwLock;
+use std::mem;
+use bytes::Bytes;
+use std::sync::Arc;
+use crate::cache::{Cache, Error, GCCallback, RCOptions};
+use std::collections::HashMap;
+use std::vec::Vec;
+use std::thread;
+use std::sync::Mutex;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use std::cmp::min;
+
+// memtable status changes from FLUSHED->MUTABLE->IMMUTABLE->FLUSHED
+#[derive(Clone, Debug, PartialEq)]
+pub enum MemTableStatus {
+    FLUSHED,    // also initial status
+    MUTABLE,
+    IMMUTABLE,
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct MemTableEntry {
+    key: Bytes,
+    val: Bytes,
+    epoch: u64, 
+    deleted: bool,
+}
+
+impl MemTableEntry {
+    pub fn new(key: Bytes, val: Bytes, epoch: u64, deleted: bool, ) -> Self {
+        MemTableEntry { key, val, epoch, deleted }
+    }
+}
+
+impl PartialOrd for MemTableEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for MemTableEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let key_comp = self.key.cmp(&other.key);
+        match key_comp {
+            Ordering::Less | Ordering::Greater => key_comp,
+            Ordering::Equal => other.epoch.cmp(&self.epoch),
+        }
+    }
+}
+
+impl PartialEq for MemTableEntry {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key && self.epoch == other.epoch
+    }
+}
+
+#[derive(Debug)]
+pub struct MemTable {
+    id: usize,
+    size_in_bytes: u64,
+    status: MemTableStatus,
+    latest_epoch: u64,
+    oldest_epoch: u64,
+    entries: OrderedSkipList<MemTableEntry>,
+}
+
+#[derive(Debug)]
+pub struct MemTableDB {
+    rc_options: RCOptions,
```

Comment:
is `rc` a fairly common Rust abbreviation? If not, I'd recommend using the full name.

Reply:
yea, rc stands for reference counting which can be confusing in this context. rc_options is range cache options. I will change the naming to
`cache_options: CacheOptions`